### PR TITLE
Move all layer tree operations to a ScrollTree struct

### PR DESCRIPTION
**Pull request by mrobinson**
This will make it easier to introduce the idea of reference frames
into WebRender. This should not change behavior.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/754)

### CODE REVIEW

**Comment by glennw** [comment]
This looks good, and I've tested scrolling on a number of different sites.

**Comment by mrobinson** [comment]
Thanks!

File: webrender/src/frame.rs [hunk 1]
```diff
@@ -334,172 +231,11 @@ impl Frame {
...
-                layer.stretch_overscroll_spring();
-            }
-
-            scrolled_a_layer = scrolled_a_layer ||
-                layer.scrolling.offset != original_layer_scroll_offset ||
-                layer.scrolling.started_bouncing_back;
-        }
-
-        scrolled_a_layer
+        self.scroll_tree.scroll(scroll_location, cursor, phase,)
```
Comment (kvark) [inline webrender/src/frame.rs]: extra comma?

File: webrender/src/frame.rs [hunk 2]
```diff
@@ -746,11 +460,11 @@ impl Frame {
 
         if level == 0 && self.frame_builder_config.enable_scrollbars {
             let scrollbar_rect = LayerRect::new(LayerPoint::zero(), LayerSize::new(10.0, 70.0));
-            context.builder.add_solid_rectangle(&scrollbar_rect,
-                                                &ClipRegion::simple(&scrollbar_rect),
-                                                &DEFAULT_SCROLLBAR_COLOR,
-                                                PrimitiveFlags::Scrollbar(self.root_scroll_layer_id.unwrap(),
-                                                                          4.0));
+            context.builder.add_solid_rectangle(
```
Comment (kvark) [inline webrender/src/frame.rs]: isn't the new formatting (see below) against the style guidelines?

File: webrender/src/scroll_tree.rs [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+            pending_scroll_offsets: HashMap::new(),
+            current_scroll_layer_id: None,
+            root_scroll_layer_id: None,
+        }
+    }
+
+    pub fn collect_layers_bouncing_back(&self)
+                                        -> HashSet<ScrollLayerId, BuildHasherDefault<FnvHasher>> {
+        let mut layers_bouncing_back = HashSet::with_hasher(Default::default());
+        for (scroll_layer_id, layer) in self.layers.iter() {
```
Comment (kvark) [inline webrender/src/scroll_tree.rs]: nit (unrelated to the PR): we could do something fancy like
```rust
self.layers.iter()
           .filter(|(_, layer)| layer.scrolling.bouncing_back)
           .map(|(id, _)| *id)
           .collect()
```

File: webrender/src/frame.rs (hunk 2 above)
Comment (mrobinson) [inline webrender/src/frame.rs]: I'm not sure, but the PrimitiveFlags::Scrollbar(self.root_scroll_layer_id.unwrap(), line wouldn't fit within the line length limit any other way.

File: webrender/src/frame.rs (hunk 1 above)
Comment (mrobinson) [inline webrender/src/frame.rs]: Yes! I will fix this in a followup.

File: webrender/src/frame.rs (hunk 2 above)
Comment (kvark) [inline webrender/src/frame.rs]: ok, I don't know what the guidelines are for this case then

**Pull request merged**